File: crates/openduo-core/src/gitlab_client.rs
use crate::auth::AuthHeaders;
use crate::config::Config;
use anyhow::Result;
use reqwest::Client;
use serde::de::DeserializeOwned;
use tracing::instrument;

#[derive(Clone)]
pub struct GitLabClient {
    client: Client,
    base_url: String,
    pat: String,
}

impl GitLabClient {
    pub fn new(config: Config) -> Self {
        let client = Client::builder()
            .use_native_tls()
            .build()
            .expect("Failed to build reqwest client");
        Self {
            client,
            base_url: config.gitlab_url.trim_end_matches('/').to_string(),
            pat: config.pat,
        }
    }

    pub fn base_url(&self) -> &str {
        &self.base_url
    }

    pub fn api_url(&self, path: &str) -> String {
        format!("{}/api/v4/{}", self.base_url, path.trim_start_matches('/'))
    }

    #[instrument(skip(self))]
    pub async fn get<T: DeserializeOwned>(&self, path: &str) -> Result<T> {
        let headers = AuthHeaders::new(&self.pat).to_header_map()?;
        let resp = self
            .client
            .get(self.api_url(path))
            .headers(headers)
            .send()
            .await?
            .error_for_status()?;
        Ok(resp.json::<T>().await?)
    }

    #[instrument(skip(self, body))]
    pub async fn post<T: DeserializeOwned>(
        &self,
        path: &str,
        body: serde_json::Value,
    ) -> Result<T> {
        let headers = AuthHeaders::new(&self.pat).to_header_map()?;
        let resp = self
            .client
            .post(self.api_url(path))
            .headers(headers)
            .json(&body)
            .send()
            .await?
            .error_for_status()?;
        Ok(resp.json::<T>().await?)
    }

    #[instrument(skip(self, body))]
    pub async fn put<T: DeserializeOwned>(&self, path: &str, body: serde_json::Value) -> Result<T> {
        let headers = AuthHeaders::new(&self.pat).to_header_map()?;
        let resp = self
            .client
            .put(self.api_url(path))
            .headers(headers)
            .json(&body)
            .send()
            .await?
            .error_for_status()?;
        Ok(resp.json::<T>().await?)
    }

    pub async fn get_raw(&self, url: &str) -> Result<reqwest::Response> {
        let headers = AuthHeaders::new(&self.pat).to_header_map()?;
        Ok(self
            .client
            .get(url)
            .headers(headers)
            .send()
            .await?
            .error_for_status()?)
    }

    pub async fn post_stream(
        &self,
        url: &str,
        body: serde_json::Value,
    ) -> Result<reqwest::Response> {
        let headers = AuthHeaders::new(&self.pat).to_header_map()?;
        Ok(self
            .client
            .post(url)
            .headers(headers)
            .json(&body)
            .send()
            .await?
            .error_for_status()?)
    }
}

File: extension/server/gitlabClient.ts
import { privateTokenHeaders } from './auth';
import { classifyError, formatCauseChain, type Config, type DebugFlags } from './config';

export class GitLabClient {
  private readonly baseUrl: string;
  private readonly pat: string;
  private readonly debug: DebugFlags;
  private readonly hasProxy: boolean;

  constructor(config: Config) {
    this.baseUrl = config.gitlabUrl.replace(/\/+$/, '');
    this.pat = config.pat;
    this.debug = config.debug;
    this.hasProxy = !!config.proxyUrl;
  }

  apiUrl(path: string): string {
    return `${this.baseUrl}/api/v4/${path.replace(/^\/+/, '')}`;
  }

  private debugLog(context: string, err: unknown): void {
    const categories = classifyError(err, this.hasProxy);
    const msg = err instanceof Error ? err.message : String(err);
    const code = (err as any)?.code ?? '';
    const stack = err instanceof Error ? err.stack ?? '' : '';
    const causeChain = formatCauseChain(err);

    for (const cat of categories) {
      if (this.debug[cat]) {
        console.log(`[${cat}] [gitlabClient] ${context}: ${msg}${code ? ` (code=${code})` : ''}`);
        if (causeChain) console.log(`[${cat}] [gitlabClient] cause chain:\n${causeChain}`);
        if (stack) console.log(`[${cat}] [gitlabClient] stack: ${stack}`);
      }
    }
    if (categories.length === 0 && this.debug.serverErrors) {
      console.log(`[serverErrors] [gitlabClient] ${context}: ${msg}`);
      if (causeChain) console.log(`[serverErrors] [gitlabClient] cause chain:\n${causeChain}`);
    }
  }

  private logApiError(method: string, url: string, status: number, body: string): void {
    if (this.debug.apiErrors) {
      console.log(`[apiErrors] [gitlabClient] ${method} ${url}: HTTP ${status}`);
      console.log(`[apiErrors] [gitlabClient] response body: ${body.slice(0, 2000)}`);
    }
  }

  async get<T = unknown>(path: string): Promise<T> {
    const url = this.apiUrl(path);
    let resp: Response;
    try {
      resp = await fetch(url, {
        headers: privateTokenHeaders(this.pat),
      });
    } catch (err) {
      this.debugLog(`GET ${url}`, err);
      throw err;
    }
    if (!resp.ok) {
      const text = await resp.text().catch(() => '');
      this.logApiError('GET', path, resp.status, text);
      throw new Error(`GitLab API GET ${path} returned ${resp.status}: ${text}`);
    }
    return resp.json() as Promise<T>;
  }

  async post<T = unknown>(path: string, body: unknown): Promise<T> {
    const url = this.apiUrl(path);
    let resp: Response;
    try {
      resp = await fetch(url, {
        method: 'POST',
        headers: privateTokenHeaders(this.pat),
        body: JSON.stringify(body),
      });
    } catch (err) {
      this.debugLog(`POST ${url}`, err);
      throw err;
    }
    if (!resp.ok) {
      const text = await resp.text().catch(() => '');
      this.logApiError('POST', path, resp.status, text);
      throw new Error(`GitLab API POST ${path} returned ${resp.status}: ${text}`);
    }
    return resp.json() as Promise<T>;
  }

  async put<T = unknown>(path: string, body: unknown): Promise<T> {
    const url = this.apiUrl(path);
    let resp: Response;
    try {
      resp = await fetch(url, {
        method: 'PUT',
        headers: privateTokenHeaders(this.pat),
        body: JSON.stringify(body),
      });
    } catch (err) {
      this.debugLog(`PUT ${url}`, err);
      throw err;
    }
    if (!resp.ok) {
      const text = await resp.text().catch(() => '');
      this.logApiError('PUT', path, resp.status, text);
      throw new Error(`GitLab API PUT ${path} returned ${resp.status}: ${text}`);
    }
    return resp.json() as Promise<T>;
  }

  async getRaw(url: string): Promise<string> {
    let resp: Response;
    try {
      resp = await fetch(url, {
        headers: privateTokenHeaders(this.pat),
      });
    } catch (err) {
      this.debugLog(`GET (raw) ${url}`, err);
      throw err;
    }
    if (!resp.ok) {
      const text = await resp.text().catch(() => '');
      this.logApiError('GET', url, resp.status, text);
      throw new Error(`GitLab API GET ${url} returned ${resp.status}: ${text}`);
    }
    return resp.text();
  }

  async delete<T = unknown>(path: string): Promise<T> {
    const url = this.apiUrl(path);
    let resp: Response;
    try {
      resp = await fetch(url, {
        method: 'DELETE',
        headers: privateTokenHeaders(this.pat),
      });
    } catch (err) {
      this.debugLog(`DELETE ${url}`, err);
      throw err;
    }
    if (!resp.ok) {
      const text = await resp.text().catch(() => '');
      this.logApiError('DELETE', path, resp.status, text);
      throw new Error(`GitLab API DELETE ${path} returned ${resp.status}: ${text}`);
    }
    // DELETE often returns 204 No Content
    if (resp.status === 204) return {} as T;
    return resp.json() as Promise<T>;
  }

  async postStream(url: string, body: unknown): Promise<Response> {
    let resp: Response;
    try {
      resp = await fetch(url, {
        method: 'POST',
        headers: privateTokenHeaders(this.pat),
        body: JSON.stringify(body),
      });
    } catch (err) {
      this.debugLog(`POST (stream) ${url}`, err);
      throw err;
    }
    if (!resp.ok) {
      const text = await resp.text().catch(() => '');
      this.logApiError('POST', url, resp.status, text);
      throw new Error(`GitLab API POST stream ${url} returned ${resp.status}: ${text}`);
    }
    return resp;
  }

  /**
   * Execute a GraphQL query against GitLab's `/api/graphql` endpoint.
   * Returns the `data` portion of the response, or throws on errors.
   */
  async graphql<T = unknown>(
    query: string,
    variables?: Record<string, unknown>,
  ): Promise<T> {
    const url = `${this.baseUrl}/api/graphql`;
    let resp: Response;
    try {
      resp = await fetch(url, {
        method: 'POST',
        headers: privateTokenHeaders(this.pat),
        body: JSON.stringify({ query, variables }),
      });
    } catch (err) {
      this.debugLog(`POST ${url} (graphql)`, err);
      throw err;
    }
    if (!resp.ok) {
      const text = await resp.text().catch(() => '');
      this.logApiError('POST', url, resp.status, text);
      throw new Error(`GitLab GraphQL returned ${resp.status}: ${text}`);
    }
    const json = (await resp.json()) as {
      data?: T;
      errors?: { message: string }[];
    };
    if (json.errors?.length) {
      if (this.debug.apiErrors) {
        console.log(`[apiErrors] [gitlabClient] GraphQL errors: ${JSON.stringify(json.errors)}`);
      }
      throw new Error(`GraphQL errors: ${json.errors.map((e) => e.message).join('; ')}`);
    }
    return json.data as T;
  }
}
